core/src/impl/Kokkos_Combined_Reducer.hpp

```diff
@@ -478,6 +496,34 @@ KOKKOS_INLINE_FUNCTION constexpr auto make_wrapped_combined_functor(
 
 template <typename FunctorType>
 using functor_has_value_t = typename FunctorType::value_type;
+
+template <typename MemberType, typename BoundaryStructType, typename Functor,
+          typename ReturnType1, typename ReturnType2, typename... ReturnTypes>
+KOKKOS_INLINE_FUNCTION void parallel_reduce_combined_reducers_impl(
+    BoundaryStructType const& boundaries, Functor const& functor,
+    ReturnType1&& returnType1, ReturnType2&& returnType2,
+    ReturnTypes&&... returnTypes) noexcept {
+  using mem_space_type = typename MemberType::scratch_memory_space;
+
+  auto combined_value = Impl::make_combined_reducer_value<mem_space_type>(
+      returnType1, returnType2, returnTypes...);
+
+  auto combined_functor = Impl::make_wrapped_combined_functor<mem_space_type>(
+      functor, returnType1, returnType2, returnTypes...);
+
+  auto combined_reducer = Impl::make_combined_reducer<mem_space_type>(
+      combined_value, returnType1, returnType2, returnTypes...);
+
+  parallel_reduce(boundaries, combined_functor, combined_reducer);
+  memory_fence();
```

Comment:
Should this rather be a `team_barrier`?

Reply:
I'll resume on this shortly.
I recall looking into a workaround to consistently call `team_barrier` for all cases after learning that a reference to `team_member` is not stored in `ThreadVectorBoundariesStruct` for a few backends because of their constructors that do not require `team_member`.

Reply:
@masterleinad I've been giving rounds of thoughts on this, and I'm thinking that the fence operations are not necessary here, both fences on line 518 and 524.

These fences were to ensure that reduced values are ready to be copied back to the original references for all reducers in a combined reducer. This is meaningful if reductions are done on device and if the result from device needs to be copied back to reducers on host (this is what the regular combined reducer does).
However, for team combined reducers, both combined reduction part and copying back to its reducer part are on device and are done in sequence.

For example, for cuda `team_reduce` that is called from `parallel_reduce` that takes in a team policy:
https://github.com/kokkos/kokkos/blob/c09dd1c94861541b2f9651474ecdbec9e074dab7/core/src/Cuda/Kokkos_Cuda_Team.hpp#L192-L204
Syncing threads is already done within the reduction. Then the result from (combined) reduction is assigned in (combined) reducer, and thus another fence afterward shouldn't be necessary to start referencing reduced values on device.

Reply:
Sounds good!

Reply:
Removed fences in [af212c0](https://github.com/kokkos/kokkos/pull/5727/commits/af212c0bc124056dc22706f18b1ac3da5c06f7ef).